File: cli/utils/display.ts
/**
 * CLI Display utilities for formatting output.
 */

/**
 * ANSI color codes.
 */
export const colors = {
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  red: "\x1b[31m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  blue: "\x1b[34m",
  magenta: "\x1b[35m",
  cyan: "\x1b[36m",
  white: "\x1b[37m",
  bgRed: "\x1b[41m",
  bgGreen: "\x1b[42m",
  bgYellow: "\x1b[43m",
  bgBlue: "\x1b[44m",
};

/**
 * Apply color to text.
 */
export function color(text: string, ...codes: string[]): string {
  return codes.join("") + text + colors.reset;
}

/**
 * Display a header box.
 */
export function header(title: string): void {
  const width = 60;
  const padding = Math.max(0, Math.floor((width - title.length - 2) / 2));
  const line = "═".repeat(width);

  console.log();
  console.log(color(`╔${line}╗`, colors.cyan));
  console.log(color(`║${" ".repeat(padding)}${title}${" ".repeat(width - padding - title.length)}║`, colors.cyan));
  console.log(color(`╚${line}╝`, colors.cyan));
  console.log();
}

/**
 * Display a section title.
 */
export function section(title: string): void {
  console.log();
  console.log(color(title, colors.bold, colors.yellow));
  console.log(color("━".repeat(title.length + 4), colors.dim));
  console.log();
}

/**
 * Display a success message.
 */
export function success(message: string): void {
  console.log(color(`✓ ${message}`, colors.green));
}

/**
 * Display an error message.
 */
export function error(message: string): void {
  console.log(color(`✗ ${message}`, colors.red));
}

/**
 * Display a warning message.
 */
export function warning(message: string): void {
  console.log(color(`⚠ ${message}`, colors.yellow));
}

/**
 * Display an info message.
 */
export function info(message: string): void {
  console.log(color(`ℹ ${message}`, colors.blue));
}

/**
 * Display a step in a process.
 */
export function step(message: string, done = false): void {
  const icon = done ? color("✓", colors.green) : color("→", colors.blue);
  console.log(`  ${icon} ${message}`);
}

/**
 * Mask an API key for display.
 */
export function maskApiKey(key: string | undefined): string {
  if (!key) return color("No configurada", colors.dim);
  if (key.length <= 10) return "***";
  return `${key.slice(0, 7)}...${key.slice(-4)}`;
}

/**
 * Display a simple table.
 */
export function table(
  headers: string[],
  rows: string[][],
  colWidths?: number[]
): void {
  // Calculate column widths
  const widths = colWidths || headers.map((h, i) => {
    const maxContent = Math.max(h.length, ...rows.map((r) => (r[i] || "").length));
    return Math.min(maxContent + 2, 30);
  });

  // Helper to pad string
  const pad = (str: string, width: number) => {
    const stripped = str.replace(/\x1b\[[0-9;]*m/g, ""); // Remove ANSI codes for length calc
    const padding = Math.max(0, width - stripped.length);
    return str + " ".repeat(padding);
  };

  // Build separator
  const sep = "─";
  const topBorder = "┌" + widths.map((w) => sep.repeat(w)).join("┬") + "┐";
  const midBorder = "├" + widths.map((w) => sep.repeat(w)).join("┼") + "┤";
  const botBorder = "└" + widths.map((w) => sep.repeat(w)).join("┴") + "┘";

  // Print table
  console.log(color(topBorder, colors.dim));

  // Headers
  const headerRow = "│" + headers.map((h, i) => pad(` ${h}`, widths[i] ?? 10)).join("│") + "│";
  console.log(color(headerRow, colors.bold));
  console.log(color(midBorder, colors.dim));

  // Rows
  for (const row of rows) {
    const rowStr = "│" + row.map((cell, i) => pad(` ${cell || ""}`, widths[i] ?? 10)).join("│") + "│";
    console.log(rowStr);
  }

  console.log(color(botBorder, colors.dim));
}

/**
 * Display a key-value list.
 */
export function keyValue(items: { key: string; value: string }[]): void {
  const maxKeyLen = Math.max(...items.map((i) => i.key.length));

  for (const item of items) {
    const key = color(item.key.padEnd(maxKeyLen), colors.dim);
    console.log(`  ${key}  ${item.value}`);
  }
}

/**
 * Display an API key box (for newly created keys).
 */
export function apiKeyBox(key: string): void {
  const width = 68;
  const line = "═".repeat(width);

  console.log();
  console.log(color(`╔${line}╗`, colors.green));
  console.log(color(`║${" ".repeat(20)}🔑 NUEVA API KEY CREADA${" ".repeat(23)}║`, colors.green));
  console.log(color(`╠${line}╣`, colors.green));
  console.log(color(`║${" ".repeat((width - key.length) / 2)}${key}${" ".repeat(Math.ceil((width - key.length) / 2))}║`, colors.green));
  console.log(color(`╠${line}╣`, colors.green));
  console.log(color(`║  ⚠️  Guarda esta key ahora. No se mostrará de nuevo!${" ".repeat(14)}║`, colors.yellow));
  console.log(color(`╚${line}╝`, colors.green));
  console.log();
}

/**
 * Status indicator.
 */
export function status(active: boolean, activeText = "Activo", inactiveText = "Inactivo"): string {
  return active
    ? color(`✓ ${activeText}`, colors.green)
    : color(`✗ ${inactiveText}`, colors.red);
}

/**
 * Checkbox indicator.
 */
export function checkbox(checked: boolean): string {
  return checked ? color("[✓]", colors.green) : color("[ ]", colors.dim);
}

File: internal/cli/display.go
package cli

import (
	"fmt"
	"strings"
	"unicode/utf8"
)

// ── ANSI color constants ──────────────────────────────────────────────────────

const (
	Reset   = "\033[0m"
	Bold    = "\033[1m"
	Dim     = "\033[2m"
	Red     = "\033[31m"
	Green   = "\033[32m"
	Yellow  = "\033[33m"
	Blue    = "\033[34m"
	Magenta = "\033[35m"
	Cyan    = "\033[36m"
	White   = "\033[37m"
)

// Color wraps s in the given ANSI color code.
func Color(color, s string) string {
	return color + s + Reset
}

// ── headers / sections ────────────────────────────────────────────────────────

// Header prints a cyan-bordered box with the given title.
func Header(title string) {
	const width = 52
	fmt.Println(Cyan + "╔" + strings.Repeat("═", width-2) + "╗" + Reset)
	inner := "  " + title
	pad := width - 2 - utf8.RuneCountInString(inner)
	if pad < 0 {
		pad = 0
	}
	fmt.Println(Cyan + "║" + Reset + inner + strings.Repeat(" ", pad) + Cyan + "║" + Reset)
	fmt.Println(Cyan + "╚" + strings.Repeat("═", width-2) + "╝" + Reset)
	fmt.Println()
}

// Section prints a yellow, underlined section heading.
func Section(title string) {
	fmt.Println(Yellow + Bold + title + Reset)
	fmt.Println(Yellow + strings.Repeat("─", utf8.RuneCountInString(title)) + Reset)
	fmt.Println()
}

// ── status lines ─────────────────────────────────────────────────────────────

func Success(msg string) { fmt.Println(Green + "  ✓ " + msg + Reset) }
func Error(msg string)   { fmt.Println(Red + "  ✗ " + msg + Reset) }
func Warning(msg string) { fmt.Println(Yellow + "  ⚠ " + msg + Reset) }
func Info(msg string)    { fmt.Println(Cyan + "  ℹ " + msg + Reset) }

// Step prints a step line; if done is true it uses a check mark.
func Step(msg string, done bool) {
	if done {
		fmt.Println(Green + "  ✓ " + msg + Reset)
	} else {
		fmt.Println(Dim + "  → " + msg + Reset)
	}
}

// Checkbox returns a colored checkbox symbol.
func Checkbox(checked bool) string {
	if checked {
		return Color(Green, "☑")
	}
	return Color(Dim, "☐")
}

// ── table ─────────────────────────────────────────────────────────────────────

// Table prints an ASCII bordered table.
// widths[i] is the inner column width (excluding padding and borders).
func Table(headers []string, rows [][]string, widths []int) {
	sep := "+" + joinSep(widths) + "+"

	fmt.Println(sep)
	fmt.Print("|")
	for i, h := range headers {
		w := colWidth(widths, i)
		fmt.Print(" " + Bold + padRight(stripANSI(h), w, h) + Reset + " |")
	}
	fmt.Println()
	fmt.Println(sep)

	for _, row := range rows {
		fmt.Print("|")
		for i, cell := range row {
			if i >= len(widths) {
				break
			}
			w := colWidth(widths, i)
			fmt.Print(" " + padRight(stripANSI(cell), w, cell) + " |")
		}
		fmt.Println()
	}
	fmt.Println(sep)
}

func joinSep(widths []int) string {
	parts := make([]string, len(widths))
	for i, w := range widths {
		parts[i] = strings.Repeat("-", w+2)
	}
	return strings.Join(parts, "+")
}

func colWidth(widths []int, i int) int {
	if i < len(widths) {
		return widths[i]
	}
	return 20
}

// padRight pads visible (ANSI-stripped) text to width w, using original to
// preserve color codes.
func padRight(visible string, w int, original string) string {
	vlen := utf8.RuneCountInString(visible)
	if vlen >= w {
		if vlen > w {
			// truncate visible
			runes := []rune(visible)
			return string(runes[:w])
		}
		return original
	}
	return original + strings.Repeat(" ", w-vlen)
}

// stripANSI removes ANSI escape codes from s to get the visible length.
func stripANSI(s string) string {
	var out strings.Builder
	i := 0
	for i < len(s) {
		if s[i] == '\033' && i+1 < len(s) && s[i+1] == '[' {
			// skip until 'm'
			j := i + 2
			for j < len(s) && s[j] != 'm' {
				j++
			}
			i = j + 1
			continue
		}
		out.WriteByte(s[i])
		i++
	}
	return out.String()
}

// ── key-value list ────────────────────────────────────────────────────────────

// KeyValue prints aligned key: value pairs.
func KeyValue(items [][2]string) {
	maxKey := 0
	for _, kv := range items {
		if n := utf8.RuneCountInString(kv[0]); n > maxKey {
			maxKey = n
		}
	}
	for _, kv := range items {
		pad := maxKey - utf8.RuneCountInString(kv[0])
		fmt.Printf("  %s%s : %s\n", kv[0], strings.Repeat(" ", pad), kv[1])
	}
}

// ── API key box ───────────────────────────────────────────────────────────────

// ApiKeyBox displays the new API key in a prominent green-bordered box.
func ApiKeyBox(key string) {
	width := len(key) + 6
	if width < 50 {
		width = 50
	}
	fmt.Println()
	fmt.Println(Green + "┌" + strings.Repeat("─", width-2) + "┐" + Reset)
	fmt.Println(Green + "│" + Reset + Bold + "  ¡GUARDA ESTA KEY — SE MUESTRA UNA SOLA VEZ!" + Reset + strings.Repeat(" ", width-2-47) + Green + "│" + Reset)
	fmt.Println(Green + "│" + Reset)
	inner := "  " + key
	pad := width - 2 - len(inner)
	if pad < 0 {
		pad = 0
	}
	fmt.Println(Green + "│" + Reset + Green + Bold + inner + Reset + strings.Repeat(" ", pad) + Green + "│" + Reset)
	fmt.Println(Green + "│" + Reset)
	fmt.Println(Green + "└" + strings.Repeat("─", width-2) + "┘" + Reset)
	fmt.Println()
}

// MaskKey shows the first 7 and last 4 chars of key, with "..." in between.
func MaskKey(key string) string {
	if len(key) < 12 {
		return key
	}
	return key[:7] + "..." + key[len(key)-4:]
}
